network/openssl_privatekey.py

```diff
@@ -0,0 +1,281 @@
+                self.remove(publickey_only=True)
+                pass
+
+    def remove(self, publickey_only=False):
+        """Remove the private key from the filesystem."""
+
+        if not publickey_only:
+            try:
+                os.remove(self.path)
+            except OSError:
+                e = get_exception()
+                if e.errno != errno.ENOENT:
+                    raise PrivateKeyError(e)
+                else:
+                    self.changed = False
+
+        if self.publickey_path:
+            try:
+                os.remove(self.publickey_path)
+            except OSError:
+                e = get_exception()
+                if e.errno != errno.ENOENT:
+                    raise PrivateKeyError(e)
+
+    def dump(self):
+        """Serialize the object into a dictionnary."""
+
+        result = {
+            'size': self.size,
+            'type': self.type,
+            'filename': self.path,
+            'changed': self.changed,
+        }
+
+        if self.publickey_path:
+            result['publickey_filename'] = self.publickey_path
+
+        return result
+        
+
+def main():
+
+    module = AnsibleModule(
+        argument_spec = dict(
+            state=dict(default='present', choices=['present', 'absent'], type='str'),
+            size=dict(default=4096, type='int'),
+            type=dict(default='RSA', choices=['RSA', 'DSA'], type='str'),
+            force=dict(default=False, type='bool'),
+            path=dict(required=True, type='path'),
+            publickey_path=dict(required=False, type='path'),
+        ),
+        supports_check_mode = True,
+    )
+
+    if not pyopenssl_found:
+        module.fail_json(msg='the python pyOpenSSL module is required')
+
+    path = module.params['path']
+    base_dir = os.path.dirname(module.params['path'])
+
+    if not os.path.isdir(base_dir):
+        module.fail_json(name=base_dir, msg='The directory %s does not exist or the file is not a directory' % base_dir)
+
+    if module.params['publickey_path']:
+        publickey_base_dir = os.path.dirname(module.params['publickey_path'])
+        if not os.path.isdir(publickey_base_dir):
+            module.fail_json(name=publickey_base_dir, msg='The directory %s does not exist or the file is not a directory' % publickey_base_dir)
+
+    private_key = PrivateKey(module)
+    if private_key.state == 'present':
+
+        if module.check_mode:
+            result = private_key.dump()
+            result['changed'] = module.params['force'] or not os.path.exists(path)
+            module.exit_json(**result)
+
+        try:
+            private_key.generate()
+        except PrivateKeyError:
+            e = get_exception()
+            module.fail_json(msg=str(e))
+    else:
+
+        if module.check_mode:
+            result = private_key.dump()
+            result['changed'] = os.path.exists(path)
+            module.exit_json(**result)
+
+        try:
+            private_key.remove()
```

Comment:
We could open and verify if that's a private key, but I do not see that as blocking if it cannot be done trivially. It is nice to protect from potential spelling errors and such, but IMHO, that would be a very uncommon error.